R/dash.R

```diff
@@ -257,17 +256,57 @@ Dash <- R6::R6Class(
         TRUE
       })
 
-      # TODO: once implemented in dash-renderer, leverage this endpoint so
-      # we can dynamically load dependencies during `_dash-update-component`
-      # https://plotly.slack.com/archives/D07PDTRK6/p1507657249000714?thread_ts=1505157408.000123&cid=D07PDTRK6
-      dash_suite <- paste0(routes_pathname_prefix, "_dash-component-suites")
+      # This endpoint supports dynamic dependency loading 
+      # during `_dash-update-component` -- for reference:
+      # https://github.com/plotly/dash/blob/1249ffbd051bfb5fdbe439612cbec7fa8fff5ab5/dash/dash.py#L488
+      # https://docs.python.org/3/library/pkgutil.html#pkgutil.get_data
+      dash_suite <- paste0(self$config$routes_pathname_prefix, "_dash-component-suites/:package_name/*")
       route$add_handler("get", dash_suite, function(request, response, keys, ...) {
-
-        response$status <- 500L
-        response$body <- "Not yet implemented"
+        # we're looking for a bit more information than returned by the routr API
+        # for the handler keys, so the URL parsing is done manually here:
+        #
+        # 1) extract the URL from the request
+        # 2) split the URL into chunks for each of the /*/ blocks
+        # 3) retrieve the script pathname by starting from the first block following
+        #    the "package" name (for the Dash deps, not the R package) name in the
+        #    URL, then capturing the rest of the pathname afterwards
+        # 4) concatenate the strings into a path, collapsing with /
+        # 5) pass the constructed path to get_package_mapping to retrieve R pkg name
+        url <- request$url
+        url_as_strings <- strsplit(gsub("http://|https://|www\\.", "", url), "/")[[1]]
+        pkg_pos <- 1 + match(keys$package_name, url_as_strings)
+        dist_path <- paste(url_as_strings[pkg_pos:length(url_as_strings)], collapse = "/")
+        
+        dep_pkg <- get_package_mapping(dist_path, 
+                                       keys$package_name,
+                                       c(private$dependencies_internal,
+                                                    private$dependencies,
+                                                    private$dependencies_user)
+                                       )
+        
+        dep_path <- system.file(dep_pkg$rpkg_path, 
+                                package = dep_pkg$rpkg_name)
```

Comment:
from a security perspective, we have to be very careful that a malicious user won't be able to construct a URL in such a way that `dep_pkg` is a non-dash-component package (enabling that user to extract code from private packages _or_ arbitrary files on the file system). It looks like this is primarily handled by checking the URLs against `private$dependencies` in `get_package_mapping`.

Reply:
Yeah, it's interesting that you mentioned this, because I realized we haven't discussed security considerations in DashR yet -- earlier today I was reviewing open issues, and saw this (completely unrelated, but on the topic of security):

https://github.com/plotly/dashR/issues/27

Reply:
it would be good if we could isolate some of this logic into a function that we could test

Reply:
I've opened a issue related to this comment, given the potential consequences in terms of application security, and its critical nature:

https://github.com/plotly/dashR/issues/55